sdk/core/azure-core/src/http/websockets/websockets_impl.hpp

```diff
@@ -0,0 +1,372 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+#include "azure/core/http/websockets/websockets.hpp"
+#include "azure/core/http/websockets/websockets_transport.hpp"
+#include "azure/core/internal/diagnostics/log.hpp"
+#include "azure/core/internal/http/pipeline.hpp"
+#include <array>
+#include <queue>
+#include <random>
+#include <shared_mutex>
+#include <thread>
+
+// Implementation of WebSocket protocol.
+namespace Azure { namespace Core { namespace Http { namespace WebSockets { namespace _detail {
+
+  // Generator for random bytes. Used in WebSocketImplementation and tests.
+  std::vector<uint8_t> GenerateRandomBytes(size_t vectorSize);
+
+  class WebSocketImplementation {
+    enum class SocketState
+    {
+      Invalid,
+      Closed,
+      Opening,
+      Open,
+      Closing,
+    };
+
+  public:
+    WebSocketImplementation(
+        Azure::Core::Url const& remoteUrl,
+        _internal::WebSocketOptions const& options);
+
+    void Open(Azure::Core::Context const& context);
+    void Close(
+        uint16_t closeStatus,
+        std::string const& closeReason,
+        Azure::Core::Context const& context);
+    void SendFrame(
+        std::string const& textFrame,
+        bool isFinalFrame,
+        Azure::Core::Context const& context);
+    void SendFrame(
+        std::vector<uint8_t> const& binaryFrame,
+        bool isFinalFrame,
+        Azure::Core::Context const& context);
+
+    std::shared_ptr<_internal::WebSocketFrame> ReceiveFrame(Azure::Core::Context const& context);
+
+    void AddHeader(std::string const& headerName, std::string const& headerValue);
+
+    std::string const& GetNegotiatedProtocol();
+    bool IsOpen() { return m_state == SocketState::Open; }
+    bool HasBuiltInWebSocketSupport();
+
+    _internal::WebSocketStatistics GetStatistics() const;
+
+  private:
+    // WebSocket opcodes.
+    enum class SocketOpcode : uint8_t
```

Comment:
SocketOptCodeType?

Reply:
This enumeration isn't defining a kind of opcode. It's the actual opcode value.

My rule is that if it's defining a kind of thing, it should be a `XxxType`. But if it's defining values of a thing, it should just be the thing (you wouln't want us to rename HttpStatusCode to HttpStatusCodeType, would you?)